[move-operator] Start having SILGen emit lexical lifetimes and teach the optimizer how to maintain lexical lifetimes until the lexical lifetime elimination.

gottesmm opened the pull request:
I am doing this so that I can use lexical lifetimes to emit diagnostics such as
the move operator diagnostics.

gottesmm committed:
[move-operator] Start having SILGen emit lexical lifetimes and teach the optimizer how to maintain lexical lifetimes until the lexical lifetime elimination.

I am doing this so that I can use lexical lifetimes to emit diagnostics such as
the move operator diagnostics.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented on include/swift/AST/SILOptions.h:
```diff
@@ -56,7 +59,7 @@ class SILOptions {
 
   /// Enable experimental support for emitting defined borrow scopes.
-  LexicalLifetimesOption LexicalLifetimes = LexicalLifetimesOption::Off;
+  LexicalLifetimesOption LexicalLifetimes = LexicalLifetimesOption::Early;
```
This is what triggers the change in behavior.

gottesmm commented on test/Interpreter/builtin_bridge_object.swift:
```diff
@@ -1,5 +1,5 @@
-// RUN: %target-run-simple-swift(-Onone -parse-stdlib -Xfrontend -enable-copy-propagation) | %FileCheck %s --check-prefixes=CHECK,CHECK-DBG
-// RUN: %target-run-simple-swift(-O -parse-stdlib -Xfrontend -enable-copy-propagation) | %FileCheck --check-prefixes=CHECK,CHECK-OPT %s
+// RUN: %target-run-simple-swift(-Onone -parse-stdlib -Xfrontend -enable-copy-propagation -Xfrontend -disable-lexical-lifetimes) | %FileCheck %s --check-prefixes=CHECK,CHECK-DBG
```
@atrick @nate-chandler I had to turn off early lexical lifetimes in certain cases where copy propagation tests were expecting different behavior. I did the disable since copy propagation is an experimental feature. We should really look into eliminating these!

atrick reviewed (approved):
Thanks for fixing the tests

gottesmm commented:
Some of the linux failures look real. I am building on my side machine.

gottesmm commented:
@swift-ci test macOS platform

gottesmm commented:
@swift-ci test windows platform

gottesmm commented:
@swift-ci test macOS platform

gottesmm commented:
Just to be complete, macOS full test passed. I am looking into the linux issues.

gottesmm committed:
[sil] Ban casting AnyObject with a guaranteed ownership forwarding checked_cast_br and fix up semantic-arc-opts given that.

The reason why I am doing this is that currently checked_cast_br of an AnyObject
may return a different value due to boxing. As an example, this can occur when
performing a checked_cast_br of a __SwiftValue(AnyHashable(Klass())).

To model this in SIL, I added to OwnershipForwardingMixin a bit of information
that states if the instruction directly forwards ownership with a default value
of true. In checked_cast_br's constructor though I specialize this behavior if
the source type is AnyObject and thus mark the checked_cast_br as not directly
forwarding its operand. If an OwnershipForwardingMixin directly forwards
ownership, one can assume that if it forwards ownership, it will always do so in
a way that ensures that each forwarded value is rc-identical to whatever result
it algebraically forwards ownership to. So in the context of checked_cast_br, it
means that the source value is rc-identical to the argument of the success and
default blocks.

I added a verifier check to CheckedCastBr that makes sure that it can never
forward guaranteed ownership and have a source type of AnyObject.

In SemanticARCOpts, I modified the code that builds extended live ranges of
owned values (*) to check if a OwnershipForwardingMixin is directly
forwarding. If it is not directly forwarding, then we treat the use just as an
unknown consuming use. This will then prevent us from converting such an owned
value to guaranteed appropriately in such a case.

I also in SILGen needed to change checked_cast_br emission in SILGenBuilder to
perform an ensurePlusOne on the input operand (converting it to +1 with an
appropriate cleanup) if the source type is an AnyObject. I found this via the
verifier check catching this behavior from SILGen when compiling libswift. This
just shows how important IR verification of invariants can be.

(*) For those unaware, SemanticARCOpts contains a model of an owned value and
all forwarding uses of it called an OwnershipLiveRange.

rdar://85710101

gottesmm commented:
@swift-ci test

gottesmm commented:
I tracked down the linux issue to https://github.com/apple/swift/pull/40287

gottesmm commented:
I disabled the two tests that are failing on Linux/windows so I can get the rest of the testing done.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
I figured out that the two remaining test failures were just due to the test relying on enable copy propagation. So I disabled lexical lifetimes on that test for now until Nate has a chance to look at that.

gottesmm commented:
@swift-ci test

gottesmm commented:
Since the bots are backed up, if all of the tests pass, then I am going to merge this despite the first commit already being merged in the other PR. I just think it will let me get it done faster without any need to retest/overuse resources.
